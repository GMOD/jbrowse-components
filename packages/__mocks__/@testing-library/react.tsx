import React, { Suspense } from 'react'
import { createTheme, ThemeProvider } from '@mui/material/styles'

const react = jest.requireActual('@testing-library/react')
const render = (args: React.ReactNode) => {
  return react.render(
    <Suspense fallback={null}>
      <ThemeProvider theme={createTheme()}>{args}</ThemeProvider>
    </Suspense>,
  )
}

module.exports = { ...react, render }
