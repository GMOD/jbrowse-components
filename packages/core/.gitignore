.DS_Store
node_modules
coverage
dist
*.log
*.swp
*.js
*.d.ts
*.js.map
./BaseFeatureWidget
./ReExports
./TextSearch
./assemblyManager
./configuration
./data_adapterts
./pluggableElementTypes
./rpc
./ui
./util
