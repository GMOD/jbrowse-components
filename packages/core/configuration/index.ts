export {
  ConfigurationSchema,
  ConfigurationReference,
} from './configurationSchema'

export type {
  AnyConfigurationSchemaType,
  AnyConfigurationModel,
  AnyConfigurationSlot,
  AnyConfigurationSlotType,
  AnyConfiguration,
} from './types'

export * from './util'
