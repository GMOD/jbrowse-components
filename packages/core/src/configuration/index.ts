export {
  ConfigurationSchema,
  ConfigurationReference,
} from './configurationSchema'

export * from './util'
