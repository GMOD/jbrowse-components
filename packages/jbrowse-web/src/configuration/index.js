import {
  isStateTreeNode,
  getPropertyMembers,
  getSnapshot,
  getType,
  isMapType,
} from 'mobx-state-tree'
import { isObservableArray, isObservableObject, isObservableMap } from 'mobx'

import {
  ConfigurationSchema,
  ConfigurationReference,
} from './configurationSchema'

function getModelConfig(tree) {
  // if this is a node
  if (isStateTreeNode(tree)) {
    let config
    if (isObservableObject(tree)) {
      let keys
      //   if it has a 'configuration' view, use that as the node instead
      //   otherwise, just recurse through it normally
      if (tree.configuration) {
        tree = tree.configuration
        keys = Object.keys(tree)
      } else {
        keys = Object.keys(getPropertyMembers(tree).properties)
      }
      config = {}
      keys.forEach(key => {
        config[key] = getModelConfig(tree[key])
      })
    } else if (isObservableArray(tree)) config = tree.map(getModelConfig)
    else if (isObservableMap(tree)) {
      config = {}
      tree.forEach((value, key) => {
        config[key] = getModelConfig(value)
      })
    }

    return config
  }
  return tree
}

/**
 * helper method for readConfObject, reads the config from a mst model
 *
 * @param {object} model instance of ConfigurationSchema
 * @param {array} slotPaths array of paths to read
 * @param {any} args extra arguments e.g. for a feature callback, will be sent to each of the slotNames
 */
function getConf(model, slotName, args) {
  if (!model.configuration)
    throw new Error(`cannot getConf on this model, it has no configuration`)
  return readConfObject(model.configuration, slotName, args)
}
/**
 * helper method, applies getConf to an array of slotNames
 *
 * @param {object} model instance of ConfigurationSchema
 * @param {array} slotNames array of config paths to read
 * @param {any} args extra arguments e.g. for a feature callback, will be sent to each of the slotNames
 */
function getConfs(model, slotNames, args) {
  return slotNames.map(slotName => getConf(model, slotName, args))
}
/**
 * helper method, applies readConfObject to an array of slotNames
 *
 * @param {object} model instance of mst model with a configuration object
 * @param {array} slotNames array of config paths to read
 * @param {any} args extra arguments e.g. for a feature callback, will be sent to each of the slotNames
 */
function readConfObjects(confObject, slotNames, args) {
  return slotNames.map(slotName => readConfObject(confObject, slotName, args))
}
/**
 * given a configuration model (an instance of a ConfigurationSchema),
 * read the configuration variable at the given path
 *
 * @param {object} model instance of ConfigurationSchema
 * @param {array} slotPaths array of paths to read
 * @param {any} args extra arguments e.g. for a feature callback, will be sent to each of the slotNames
 */
function readConfObject(confObject, slotPath, args) {
  if (!slotPath) return getSnapshot(confObject)
  if (typeof slotPath === 'string') {
    let slot = confObject[slotPath]
    // check for the subconf being a map if we don't find it immediately
    if (
      !slot &&
      isStateTreeNode(confObject) &&
      isMapType(getType(confObject))
    ) {
      slot = confObject.get(slotName)
    }
    if (!slot) {
      return undefined
      // if we want to be very strict about config slots, we could uncomment the below
      // instead of returning undefine
      //
      // const modelType = getType(model)
      // const schemaType = model.configuration && getType(model.configuration)
      // throw new Error(
      //   `no slot "${slotName}" found in ${modelType.name} configuration (${
      //     schemaType.name
      //   })`,
      // )
    }
    if (slot.func) {
      const appliedFunc = slot.func.apply(null, args)
      if (isStateTreeNode(appliedFunc)) return getSnapshot(appliedFunc)
      return appliedFunc
    }
    if (isStateTreeNode(slot)) return getSnapshot(slot)
    return slot
  }

  const slotName = slotPath[0]
  if (slotPath.length > 1) {
    const newPath = slotPath.slice(1)
    let subConf = confObject[slotName]
    // check for the subconf being a map if we don't find it immediately
    if (
      !subConf &&
      isStateTreeNode(confObject) &&
      isMapType(getType(confObject))
    ) {
      subConf = confObject.get(slotName)
    }
    if (!subConf) {
      return undefined
    }
    return readConfObject(subConf, newPath, args)
  }
  return readConfObject(confObject, slotName, args)
}

export {
  ConfigurationSchema,
  ConfigurationReference,
  getModelConfig,
  getConf,
  getConfs,
  readConfObject,
  readConfObjects,
}
