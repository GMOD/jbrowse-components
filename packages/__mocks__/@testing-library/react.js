import { StylesProvider } from '@material-ui/core/styles'
import React from 'react'

const react = jest.requireActual('@testing-library/react')
const generateClassName = (rule, styleSheet) =>
  `${styleSheet.options.classNamePrefix}-${rule.key}`
const render = args => {
  return react.render(
    <StylesProvider generateClassName={generateClassName}>
      {args}
    </StylesProvider>,
  )
}

module.exports = { ...react, render }
