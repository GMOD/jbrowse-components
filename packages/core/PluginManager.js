import React from 'react'
import * as mst from 'mobx-state-tree'

import RendererType from './pluggableElementTypes/renderers/RendererType'
import AdapterType from './pluggableElementTypes/AdapterType'
import TrackType from './pluggableElementTypes/TrackType'
import ViewType from './pluggableElementTypes/ViewType'
import DrawerWidgetType from './pluggableElementTypes/DrawerWidgetType'
import MenuBarType from './pluggableElementTypes/MenuBarType'
import ConnectionType from './pluggableElementTypes/ConnectionType'

import { ConfigurationSchema } from './configuration'

// little helper class that keeps groups of callbacks that are
// then run in a specified order by group
class PhasedScheduler {
  phaseCallbacks = {}

  constructor(...phaseOrder) {
    this.phaseOrder = phaseOrder
  }

  add(phase, callback) {
    if (!this.phaseOrder.includes(phase)) {
      throw new Error(`unknown phase ${phase}`)
    }
    if (!this.phaseCallbacks[phase]) this.phaseCallbacks[phase] = []
    this.phaseCallbacks[phase].push(callback)
  }

  run() {
    this.phaseOrder.forEach(phaseName => {
      if (this.phaseCallbacks[phaseName]) {
        this.phaseCallbacks[phaseName].forEach(callback => callback())
      }
    })
  }
}

export default class PluginManager {
  plugins = []

  elementTypes = {}

  elementCreationSchedule = new PhasedScheduler(
    'renderer',
    'adapter',
    'track',
    'connection',
    'view',
    'drawer widget',
    'menu bar',
  )

  typeBaseClasses = {
    renderer: RendererType,
    adapter: AdapterType,
    track: TrackType,
    connection: ConnectionType,
    view: ViewType,
    'drawer widget': DrawerWidgetType,
    'menu bar': MenuBarType,
  }

  static lib = { 'mobx-state-tree': mst, React }

  constructor(initialPlugins = []) {
    this.lib = PluginManager.lib

    this.getRendererType = this.getElementType.bind(this, 'renderer')
    this.getAdapterType = this.getElementType.bind(this, 'adapter')
    this.getTrackType = this.getElementType.bind(this, 'track')
    this.getViewType = this.getElementType.bind(this, 'view')
    this.getDrawerWidgetType = this.getElementType.bind(this, 'drawer widget')
    this.getMenuBarType = this.getElementType.bind(this, 'menu bar')
    this.getConnectionType = this.getElementType.bind(this, 'connection')
    this.addRendererType = this.addElementType.bind(this, 'renderer')
    this.addAdapterType = this.addElementType.bind(this, 'adapter')
    this.addTrackType = this.addElementType.bind(this, 'track')
    this.addViewType = this.addElementType.bind(this, 'view')
    this.addDrawerWidgetType = this.addElementType.bind(this, 'drawer widget')
    this.addMenuBarType = this.addElementType.bind(this, 'menu bar')
    this.addConnectionType = this.addElementType.bind(this, 'connection')

    // add all the initial plugins
    initialPlugins.forEach(plugin => {
      this.addPlugin(plugin)
    })
  }

  addPlugin(plugin) {
    if (this.configured) {
      throw new Error('JBrowse already configured, cannot add plugins')
    }
    if (this.plugins.includes(plugin)) {
      throw new Error('plugin already installed')
    }
    // if (!plugin.install) console.error(plugin)
    plugin.install(this)
    this.plugins.push(plugin)
    return this
  }

  /** get a MST type for the union of all specified pluggable MST types */
  pluggableMstType(
    typeGroup,
    fieldName,
    fallback = mst.types.maybe(mst.types.null),
  ) {
    const pluggableTypes = this.getElementTypeMembers(typeGroup, fieldName)
    // try to smooth over the case when no types are registered, mostly encountered in tests
    if (pluggableTypes.length === 0) return fallback
    return mst.types.union(...pluggableTypes)
  }

  /** get a MST type for the union of all specified pluggable config schemas */
  pluggableConfigSchemaType(
    typeGroup,
    fieldName = 'configSchema',
    fallback = ConfigurationSchema('Null', {}),
  ) {
    return this.pluggableMstType(typeGroup, fieldName, fallback)
  }

  addElementType(groupName, creationCallback) {
    if (typeof creationCallback !== 'function') {
      throw new Error('must provide a callback function')
    }
    const typeBaseClass = this.typeBaseClasses[groupName]
    if (!typeBaseClass) {
      throw new Error(`unknown pluggable element type ${groupName}, cannot add`)
    }
    if (!this.elementTypes[groupName]) this.elementTypes[groupName] = {}

    this.elementCreationSchedule.add(groupName, () => {
      const element = creationCallback(this)
      if (
        groupName === 'adapter' &&
        !element.AdapterClass.capabilities.length
      ) {
        throw new Error(
          `Adapter ${
            element.AdapterClass.name
          } must provide a static property "capabilities" that has at least one entry. See BaseAdapter for an example.`,
        )
      }
      if (this.elementTypes[groupName][element.name]) {
        throw new Error(
          `${groupName} ${
            element.name
          } already registered, cannot register it again`,
        )
      }

      this.elementTypes[groupName][element.name] = element
    })

    return this
  }

  getElementType(groupName, typeName) {
    if (!typeName) {
      throw new Error(`invalid type name "${typeName}"`)
    }
    return (this.elementTypes[groupName] || {})[typeName]
  }

  getElementTypesInGroup(groupName) {
    return Object.values(this.elementTypes[groupName] || {})
  }

  getElementTypeMembers(groupName, memberName) {
    return this.getElementTypesInGroup(groupName)
      .map(t => t[memberName])
      .filter(m => !!m)
  }

  configure() {
    if (this.configured) throw new Error('already configured')

    // run the creation callbacks for each element type in order.
    // see elementCreationSchedule above for the creation order
    this.elementCreationSchedule.run()
    delete this.elementCreationSchedule

    this.plugins.forEach(plugin => plugin.configure())

    this.configured = true

    // console.log(JSON.stringify(getSnapshot(model)))

    return this
  }
}
