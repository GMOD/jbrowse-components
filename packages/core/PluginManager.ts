import {
  types,
  IAnyType,
  IAnyModelType,
  isModelType,
  isType,
} from 'mobx-state-tree'

import PluggableElementBase from './pluggableElementTypes/PluggableElementBase'
import RendererType from './pluggableElementTypes/renderers/RendererType'
import AdapterType from './pluggableElementTypes/AdapterType'
import TrackType from './pluggableElementTypes/TrackType'
import DisplayType from './pluggableElementTypes/DisplayType'
import ViewType from './pluggableElementTypes/ViewType'
import WidgetType from './pluggableElementTypes/WidgetType'
import ConnectionType from './pluggableElementTypes/ConnectionType'
import RpcMethodType from './pluggableElementTypes/RpcMethodType'

import {
  ConfigurationSchema,
  isBareConfigurationSchemaType,
} from './configuration'

import Plugin from './Plugin'
import ReExports from './ReExports'

import {
  PluggableElementType,
  PluggableElementMember,
} from './pluggableElementTypes'
import { AnyConfigurationSchemaType } from './configuration/configurationSchema'
import { AbstractRootModel } from './util'
import CorePlugin from './CorePlugin'

/** little helper class that keeps groups of callbacks that are
then run in a specified order by group */
class PhasedScheduler<PhaseName extends string> {
  phaseCallbacks = new Map<PhaseName, Function[]>()

  phaseOrder: PhaseName[] = []

  constructor(...phaseOrder: PhaseName[]) {
    this.phaseOrder = phaseOrder
  }

  add(phase: PhaseName, callback: Function) {
    if (!this.phaseOrder.includes(phase)) {
      throw new Error(`unknown phase ${phase}`)
    }
    let phaseCallbacks = this.phaseCallbacks.get(phase)
    if (!phaseCallbacks) {
      phaseCallbacks = []
      this.phaseCallbacks.set(phase, phaseCallbacks)
    }
    phaseCallbacks.push(callback)
  }

  run() {
    this.phaseOrder.forEach(phaseName => {
      const phaseCallbacks = this.phaseCallbacks.get(phaseName)
      if (phaseCallbacks) {
        phaseCallbacks.forEach(callback => callback())
      }
    })
  }
}

type PluggableElementTypeGroup =
  | 'renderer'
  | 'adapter'
  | 'display'
  | 'track'
  | 'connection'
  | 'view'
  | 'widget'
  | 'rpc method'

/** internal class that holds the info for a certain element type */
class TypeRecord<ElementClass extends PluggableElementBase> {
  registeredTypes: { [name: string]: ElementClass } = {}

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  baseClass: { new (...args: any[]): ElementClass }

  typeName: string

  constructor(
    typeName: string,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    elementType: { new (...args: any[]): ElementClass },
  ) {
    this.typeName = typeName
    this.baseClass = elementType
  }

  add(name: string, t: ElementClass) {
    this.registeredTypes[name] = t
  }

  has(name: string) {
    return name in this.registeredTypes
  }

  get(name: string) {
    if (!this.has(name))
      throw new Error(
        `${this.typeName} '${name}' not found, perhaps its plugin is not loaded or its plugin has not added it.`,
      )
    return this.registeredTypes[name]
  }

  all() {
    return Object.values(this.registeredTypes)
  }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type AnyFunction = (...args: any) => any

export default class PluginManager {
  plugins: Plugin[] = []

  elementCreationSchedule = new PhasedScheduler<PluggableElementTypeGroup>(
    'renderer',
    'adapter',
    'display',
    'track',
    'connection',
    'view',
    'widget',
    'rpc method',
  )

  rendererTypes = new TypeRecord('RendererType', RendererType)

  adapterTypes = new TypeRecord('AdapterType', AdapterType)

  trackTypes = new TypeRecord('TrackType', TrackType)

  displayTypes = new TypeRecord('DisplayType', DisplayType)

  connectionTypes = new TypeRecord('ConnectionType', ConnectionType)

  viewTypes = new TypeRecord('ViewType', ViewType)

  widgetTypes = new TypeRecord('WidgetType', WidgetType)

  rpcMethods = new TypeRecord('RpcMethodType', RpcMethodType)

  configured = false

  rootModel?: AbstractRootModel

  constructor(initialPlugins: Plugin[] = []) {
    // add the core plugin
    this.addPlugin(new CorePlugin())

    // add all the initial plugins
    initialPlugins.forEach(plugin => {
      this.addPlugin(plugin)
    })
  }

  addPlugin(plugin: Plugin) {
    if (this.configured) {
      throw new Error('JBrowse already configured, cannot add plugins')
    }
    if (this.plugins.includes(plugin)) {
      throw new Error('plugin already installed')
    }
    plugin.install(this)
    this.plugins.push(plugin)
    return this
  }

  getPlugin(name: string) {
    return this.plugins.find(p => p.name === name)
  }

  hasPlugin(name: string) {
    return this.getPlugin(name) !== undefined
  }

  createPluggableElements() {
    // run the creation callbacks for each element type in order.
    // see elementCreationSchedule above for the creation order
    this.elementCreationSchedule.run()
    delete this.elementCreationSchedule
    return this
  }

  setRootModel(rootModel: AbstractRootModel) {
    this.rootModel = rootModel
  }

  configure() {
    if (this.configured) throw new Error('already configured')

    this.plugins.forEach(plugin => plugin.configure(this))

    this.configured = true

    // console.log(JSON.stringify(getSnapshot(model)))

    return this
  }

  getElementTypeRecord(
    groupName: PluggableElementTypeGroup,
  ): TypeRecord<PluggableElementBase> {
    switch (groupName) {
      case 'adapter':
        return this.adapterTypes
      case 'connection':
        return this.connectionTypes
      case 'widget':
        return this.widgetTypes
      case 'renderer':
        return this.rendererTypes
      case 'display':
        return this.displayTypes
      case 'track':
        return this.trackTypes
      case 'view':
        return this.viewTypes
      case 'rpc method':
        return this.rpcMethods
      default:
        throw new Error(`invalid element type '${groupName}'`)
    }
  }

  addElementType(
    groupName: PluggableElementTypeGroup,
    creationCallback: (pluginManager: PluginManager) => PluggableElementType,
  ) {
    if (typeof creationCallback !== 'function') {
      throw new Error(
        'must provide a callback function that returns the new type object',
      )
    }
    const typeRecord = this.getElementTypeRecord(groupName)

    this.elementCreationSchedule.add(groupName, () => {
      const newElement = creationCallback(this)
      if (!newElement.name)
        throw new Error(`cannot add a ${groupName} with no name`)

      if (typeRecord.has(newElement.name)) {
        throw new Error(
          `${groupName} ${newElement.name} already registered, cannot register it again`,
        )
      }

      typeRecord.add(newElement.name, newElement)
    })

    return this
  }

  getElementType(groupName: PluggableElementTypeGroup, typeName: string) {
    const typeRecord = this.getElementTypeRecord(groupName)
    return typeRecord.get(typeName)
  }

  getElementTypesInGroup(groupName: PluggableElementTypeGroup) {
    const typeRecord = this.getElementTypeRecord(groupName)
    return typeRecord.all()
  }

  /** get a MST type for the union of all specified pluggable MST types */
  pluggableMstType(
    typeGroup: PluggableElementTypeGroup,
    fieldName: PluggableElementMember,
    fallback: IAnyType = types.maybe(types.null),
  ) {
    const pluggableTypes: IAnyModelType[] = []
    this.getElementTypeRecord(typeGroup)
      .all()
      .forEach(t => {
        const thing = t[fieldName]
        if (isType(thing) && isModelType(thing)) {
          pluggableTypes.push(thing)
        }
      })
    // try to smooth over the case when no types are registered, mostly encountered in tests
    if (pluggableTypes.length === 0) {
      console.warn(
        `No JBrowse pluggable types found matching ('${typeGroup}','${fieldName}')`,
      )
      return fallback
    }
    return types.union(...pluggableTypes)
  }

  /** get a MST type for the union of all specified pluggable config schemas */
  pluggableConfigSchemaType(
    typeGroup: PluggableElementTypeGroup,
    fieldName: PluggableElementMember = 'configSchema',
  ) {
    const pluggableTypes: AnyConfigurationSchemaType[] = []
    this.getElementTypeRecord(typeGroup)
      .all()
      .forEach(t => {
        const thing = t[fieldName]
        if (isBareConfigurationSchemaType(thing)) {
          pluggableTypes.push(thing)
        }
      })
    if (pluggableTypes.length === 0)
      pluggableTypes.push(ConfigurationSchema('Null', {}))
    return types.union(...pluggableTypes)
  }

  jbrequireCache = new Map()

  lib = ReExports

  load = <FTYPE extends AnyFunction>(lib: FTYPE): ReturnType<FTYPE> => {
    if (!this.jbrequireCache.has(lib)) this.jbrequireCache.set(lib, lib(this))
    return this.jbrequireCache.get(lib)
  }

  /**
   * Get the re-exported version of the given package name.
   * Throws an error if the package is not re-exported by the plugin manager.
   *
   * @returns the library's default export
   */
  jbrequire = (
    lib: keyof typeof ReExports | AnyFunction | { default: AnyFunction },
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ): any => {
    if (typeof lib === 'string') {
      const pack = this.lib[lib]
      if (!pack) {
        throw new TypeError(
          `No jbrequire re-export defined for package '${lib}'. If this package must be shared between plugins, add it to ReExports.js. If it does not need to be shared, just import it normally.`,
        )
      }
      return pack
    }

    if (typeof lib === 'function') {
      return this.load(lib)
    }

    if (lib.default) return this.jbrequire(lib.default)

    throw new TypeError(
      'lib passed to jbrequire must be either a string or a function',
    )
  }

  getRendererType(typeName: string): RendererType {
    return this.rendererTypes.get(typeName)
  }

  getRendererTypes(): RendererType[] {
    return this.rendererTypes.all()
  }

  getAdapterType(typeName: string): AdapterType {
    return this.adapterTypes.get(typeName)
  }

  getTrackType(typeName: string): TrackType {
    return this.trackTypes.get(typeName)
  }

  getDisplayType(typeName: string): DisplayType {
    return this.displayTypes.get(typeName)
  }

  getViewType(typeName: string): ViewType {
    return this.viewTypes.get(typeName)
  }

  getWidgetType(typeName: string): WidgetType {
    return this.widgetTypes.get(typeName)
  }

  getConnectionType(typeName: string): ConnectionType {
    return this.connectionTypes.get(typeName)
  }

  getRpcMethodType(methodName: string): RpcMethodType {
    return this.rpcMethods.get(methodName)
  }

  addRendererType(
    creationCallback: (pluginManager: PluginManager) => RendererType,
  ): this {
    return this.addElementType('renderer', creationCallback)
  }

  addAdapterType(
    creationCallback: (pluginManager: PluginManager) => AdapterType,
  ): this {
    return this.addElementType('adapter', creationCallback)
  }

  addTrackType(
    creationCallback: (pluginManager: PluginManager) => TrackType,
  ): this {
    // Goes through the already-created displays and registers the ones that
    // specify this track type
    const callback = () => {
      const track = creationCallback(this)
      ;(this.getElementTypesInGroup('display') as DisplayType[]).forEach(
        display => {
          if (
            display.trackType === track.name &&
            // track may have already added the displayType in its creationCallback
            !track.displayTypes.includes(display)
          ) {
            track.addDisplayType(display)
          }
        },
      )
      return track
    }
    return this.addElementType('track', callback)
  }

  addDisplayType(
    creationCallback: (pluginManager: PluginManager) => DisplayType,
  ): this {
    return this.addElementType('display', creationCallback)
  }

  addViewType(
    creationCallback: (pluginManager: PluginManager) => ViewType,
  ): this {
    const callback = () => {
      const newView = creationCallback(this)
      ;(this.getElementTypesInGroup('display') as DisplayType[]).forEach(
        display => {
          if (
            display.viewType === newView.name &&
            // view may have already added the displayType in its creationCallback
            !newView.displayTypes.includes(display)
          ) {
            newView.addDisplayType(display)
          }
        },
      )
      return newView
    }
    return this.addElementType('view', callback)
  }

  addWidgetType(
    creationCallback: (pluginManager: PluginManager) => WidgetType,
  ): this {
    return this.addElementType('widget', creationCallback)
  }

  addConnectionType(
    creationCallback: (pluginManager: PluginManager) => ConnectionType,
  ): this {
    return this.addElementType('connection', creationCallback)
  }

  addRpcMethod(
    creationCallback: (pluginManager: PluginManager) => RpcMethodType,
  ): this {
    return this.addElementType('rpc method', creationCallback)
  }
}
