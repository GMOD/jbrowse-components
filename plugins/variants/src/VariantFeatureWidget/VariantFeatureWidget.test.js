import React from 'react'
import { render } from '@testing-library/react'
import { types } from 'mobx-state-tree'
import { ConfigurationSchema } from '@jbrowse/core/configuration'
import PluginManager from '@jbrowse/core/PluginManager'
import { stateModelFactory } from '.'

import VariantFeatureDetails from './VariantFeatureWidget'

describe('VariantTrack widget', () => {
  it('renders with just the required model elements', () => {
    console.warn = jest.fn()
    const pluginManager = new PluginManager([])
    const Session = types.model({
      rpcManager: types.optional(types.frozen(), {}),
      pluginManager: types.optional(types.frozen(), {}),
      configuration: ConfigurationSchema('test', {}),
      widget: stateModelFactory(pluginManager),
    })
    const model = Session.create({
      widget: { type: 'VariantFeatureWidget' },
    })
    model.widget.setFeatureData({
      refName: 'ctgA',
      start: 176,
      end: 177,
      name: 'rs123',
      REF: 'A',
      ALT: ['<TRA>'],
      QUAL: 10.4,
      INFO: {
        MQ: 5,
      },
    })

    const { container } = render(<VariantFeatureDetails model={model.widget} />)
    expect(container.firstChild).toMatchSnapshot()
  })
})
