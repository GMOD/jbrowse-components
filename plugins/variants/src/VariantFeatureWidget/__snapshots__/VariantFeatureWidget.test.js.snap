// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VariantTrack widget renders with just the required model elements 1`] = `
<div
  class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
  data-testid="variant-side-drawer"
>
  <div
    class="MuiPaper-root makeStyles-paperRoot MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="MuiPaper-root MuiAccordion-root MuiAccordion-expanded MuiAccordion-rounded MuiPaper-elevation1 MuiPaper-rounded"
      style="margin-top: 4px;"
    >
      <div
        aria-disabled="false"
        aria-expanded="true"
        class="MuiButtonBase-root MuiAccordionSummary-root MuiAccordionSummary-expanded"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiAccordionSummary-content MuiAccordionSummary-expanded"
        >
          <span
            class="MuiTypography-root MuiTypography-button"
          >
             
            Primary data
          </span>
        </div>
        <div
          aria-disabled="false"
          aria-hidden="true"
          class="MuiButtonBase-root MuiIconButton-root MuiAccordionSummary-expandIcon MuiAccordionSummary-expanded MuiIconButton-edgeEnd"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root makeStyles-expandIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </div>
      </div>
      <div
        class="MuiCollapse-container MuiCollapse-entered"
        style="min-height: 0px;"
      >
        <div
          class="MuiCollapse-wrapper"
        >
          <div
            class="MuiCollapse-wrapperInner"
          >
            <div
              role="region"
            >
              <div
                class="MuiAccordionDetails-root makeStyles-expansionPanelDetails"
              >
                <div
                  class="makeStyles-field"
                >
                  <div
                    class="makeStyles-fieldName"
                  >
                    Position
                  </div>
                  <div
                    class="makeStyles-fieldValue"
                  >
                    <div>
                      ctgA:177..177 
                    </div>
                  </div>
                </div>
                <div
                  class="makeStyles-field"
                >
                  <div
                    class="makeStyles-fieldName"
                  >
                    Name
                  </div>
                  <div
                    class="makeStyles-fieldValue"
                  >
                    <div>
                      rs123
                    </div>
                  </div>
                </div>
                <div
                  class="makeStyles-field"
                >
                  <div
                    class="makeStyles-fieldName"
                  >
                    Length
                  </div>
                  <div
                    class="makeStyles-fieldValue"
                  >
                    <div>
                      1
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <hr
      class="MuiDivider-root"
    />
    <div
      class="MuiPaper-root MuiAccordion-root MuiAccordion-expanded MuiAccordion-rounded MuiPaper-elevation1 MuiPaper-rounded"
      style="margin-top: 4px;"
    >
      <div
        aria-disabled="false"
        aria-expanded="true"
        class="MuiButtonBase-root MuiAccordionSummary-root MuiAccordionSummary-expanded"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiAccordionSummary-content MuiAccordionSummary-expanded"
        >
          <span
            class="MuiTypography-root MuiTypography-button"
          >
             
            Attributes
          </span>
        </div>
        <div
          aria-disabled="false"
          aria-hidden="true"
          class="MuiButtonBase-root MuiIconButton-root MuiAccordionSummary-expandIcon MuiAccordionSummary-expanded MuiIconButton-edgeEnd"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root makeStyles-expandIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </div>
      </div>
      <div
        class="MuiCollapse-container MuiCollapse-entered"
        style="min-height: 0px;"
      >
        <div
          class="MuiCollapse-wrapper"
        >
          <div
            class="MuiCollapse-wrapperInner"
          >
            <div
              role="region"
            >
              <div
                class="MuiAccordionDetails-root makeStyles-expansionPanelDetails"
              >
                <div
                  style="display: flex;"
                >
                  <div
                    class="makeStyles-fieldName"
                    title="reference base(s): Each base must be one of A,C,G,T,N (case insensitive)."
                  >
                    REF
                  </div>
                  <div
                    class="makeStyles-fieldValue"
                  >
                    <div>
                      A
                    </div>
                  </div>
                </div>
                <div
                  style="display: flex;"
                >
                  <div
                    class="makeStyles-fieldName"
                    title=" alternate base(s): Comma-separated list of alternate non-reference alleles"
                  >
                    ALT
                  </div>
                  <div
                    class="makeStyles-fieldValue"
                  >
                    <div>
                      &lt;TRA&gt;
                    </div>
                  </div>
                </div>
                <div
                  style="display: flex;"
                >
                  <div
                    class="makeStyles-fieldName"
                    title="quality: Phred-scaled quality score for the assertion made in ALT"
                  >
                    QUAL
                  </div>
                  <div
                    class="makeStyles-fieldValue"
                  >
                    <div>
                      10.4
                    </div>
                  </div>
                </div>
                <div
                  style="display: flex;"
                >
                  <div
                    class="makeStyles-fieldName"
                  >
                    MQ
                  </div>
                  <div
                    class="makeStyles-fieldValue"
                  >
                    <div>
                      5
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <hr
    class="MuiDivider-root"
  />
</div>
`;
